impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java

```diff
@@ -272,25 +333,17 @@ public void close() throws Exception {
     }
 
     @Override
-    public int getNumOfBrokers() {
-        return clusterConfig.getBrokersNum();
+    public synchronized int getNumOfBrokers() {
+        return servers.size();
     }
 
-    private void releaseAllPorts() {
-        releasePorts(controllerPorts);
-        releasePorts(interBrokerPorts);
-        releasePorts(anonPorts);
-        releasePorts(externalPorts);
-    }
-
-    private void releasePorts(List<ServerSocket> ports) {
-        ports.forEach(serverSocket -> {
-            try {
-                serverSocket.close();
-            }
-            catch (IOException e) {
-                LOGGER.log(System.Logger.Level.WARNING, "failed to close socket: {0} due to: {1}", serverSocket, e.getMessage(), e);
-            }
-        });
+    @Override
+    public synchronized EndpointPair getEndpointPair(Listener listener, int nodeId) {
+        var portMap = ports.get(listener);
+        if (portMap == null) {
+            throw new IllegalStateException("Unexpected value: " + listener);
+        }
+        var port = portMap.get(nodeId);
+        return EndpointPair.builder().bind(new Endpoint("0.0.0.0", port.getLocalPort())).connect(new Endpoint("localhost", port.getLocalPort())).build();
```

Comment:
This makes me wonder if it should be the EndPoints we keep in the nested map of the ports map.

Reply:
Yeah, agreed,  I'll have a poke at that and see how it looks.

Reply:
With Rob's changes, I'm reached the 'that's good enough' point on this PR.  I'm inclined not to action this now, unless others feel strongly.